### PR TITLE
Blog post, outlines KubeVirt's usage k8s extension features

### PR DESCRIPTION
**Pull request by davidvossel**
This blog post outlines how KubeVirt uses CRDs and other k8 extension mechanisms to provide our feature set.

While I do briefly explain KubeVirt for context, the goal here is to provide a real world example of the decision making process projects like ours go through when looking at the k8s tool set. We're doing our best to extend k8s for virtualized workloads in a way that integrates seamlessly with existing k8s clusters. This post provides insight into what k8s features we're using to do this and the thought process that went into our decision making.

### CODE REVIEW

File: content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+Unfortunately in Kubernetes 1.10, CRDs do not have support for multiple versions. This means when we want to progress a CRD from kubevirt.io/v1alpha1 to kubevirt.io/v1beta1, the only path currently available to us is to backup our CRD objects, delete the registered CRD from Kubernetes, register a new CRD with the updated version, convert the backed up CRD objects to the new version, and finally post the migrated CRD objects back to the cluster.
+
+That strategy is not a viable option for us.
+
+Luckily, while we don’t currently have a complete solution in place to handle CRDs with multiple versions, there is [work underway to rectify this issue in Kubernetes](https://github.com/kubernetes/features/issues/544). Looking at the Kubernetes 1.11 release notes, the [initial steps](https://github.com/kubernetes/kubernetes/pull/63830) have already been taken to kick off this feature. CRDs with multiple versions will be supported soon and we look forward to taking advantage of that feature as soon as it lands.
```
Comment (jberkus) [inline content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md]: This pretty much says that the CRD versioning in 1.11 isn't complete for Kubevirt's use case.  Can you expand on that?  What work remains to be done?  The feature in 1.11 is marked as a complete implementation.

File: content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
+---
+layout: blog
+title:  'KubeVirt: Extending Kubernetes with CRDs for Virtualized Workloads'
+date:   2018-07-09
+---
+
+**Author**: David Vossel (Red Hat)
+
```
Comment (jberkus) [inline content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md]: If we use this blog post for the 5/6 days of 1.11, we'll want to add a paragraph here referencing the 1.11 features, namely subresources and versions.

**Comment by jberkus** [comment]
Hmmm.

I thought there was more 1.11 stuff in this blog post than there actually is.  So it's not going to help us with 1.11; we should hold this post until after we're done with the 1.11 series.

File: content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md (hunk 1 above)
Comment (davidvossel) [reply content/en/blog/_posts/2018-07-09-kubevirt-crds-for-virtualization.md]: > This pretty much says that the CRD versioning in 1.11 isn't complete for Kubevirt's use case. Can you expand on that? What work remains to be done? The feature in 1.11 is marked as a complete implementation.

There's no conversion between CRD versions. It's also strange that all the versions share the same OpenAPIv3 validation. It's like we've created a way to progress an api from alpha->beta->release only if that api contains no changes.

The value of the current implementation seems limited, but it's definitely a first step in the right direction. 

I'll update the doc to reflect my thoughts here.

**Comment by davidvossel** [comment]
@kbarnard10 very close. I need to make a few adjustments and it should be good to go. I should be able to finish that last bit up this afternoon.

**Comment by davidvossel** [comment]
@kbarnard10 @jberkus I updated the blog post. From my perspective it is ready to go. Let me know if there's anything else I need to address. Thanks!

**Comment by neolit123** [comment]
i will do a quick review. saw a couple of nits.

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+layout: blog
+title:  'KubeVirt: Extending Kubernetes with CRDs for Virtualized Workloads'
+date:   2018-07-27
+---
+
+**Author**: David Vossel (Red Hat)
+
+## What is KubeVirt?
+
+[KubeVirt](https://github.com/kubevirt/kubevirt) is a Kubernetes addon that provides users the ability to schedule traditional virtual machine workloads side by side with container workloads. Through the use of [Custom Resource Definitions](https://Kubernetes.io/docs/concepts/extend-Kubernetes/api-extension/custom-resources/) (CRDs) and other Kubernetes features, KubeVirt seamlessly extends existing Kubernetes clusters to provide a set of virtualization apis that can be used to manage virtual machines.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `apis` -> `APIs`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+## Why Use CRDs Over an Aggregated API Server?
+
+Back in the middle of 2017, those of us working on KubeVirt were at a crossroads.  We had to make a decision whether or not to extend Kubernetes using an aggregated API server or to make use of the new Custom Resource Definitions (CRDs) feature.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: extra space before `We`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+We weren’t okay with this tradeoff. Our goal wasn’t to just extend Kubernetes to run virtualization workloads, it was to do it in the most seamless and effortless way possible. We felt that the added complexity involved with an aggregated API server sacrificed the part of the user experience involved with installing and operating KubeVirt.
+ 
+**Ultimately we chose to go with CRDs and trust that the Kubernetes ecosystem would grow with us to meet the needs of our use case.** Our bets were well placed. At this point there are either solutions in place or solutions under discussion that solve every feature gap we encountered back in 2017 when were evaluating CRDs vs an aggregated API server. 
+
+## Building Layered “Kubernetes like” APIs with CRDs
+
+We designed KubeVirt’s api to follow the same patterns users are already familiar with in the Kubernetes core api.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `api` -> `API`.
in a couple of places in this sentence.

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+In KubeVirt, this concept of layering controllers is at the very center of our design. The KubeVirt VirtualMachineInstance (VMI) object is the lowest level unit at the very bottom of the KubeVirt stack. Similar in concept to a Pod, a VMI represents a single mortal virtualized workload that executes once until completion (powered off).
+
+Layered on top of VMIs we have a workload controller called a VirtualMachine (VM). The VM controller is where we really begin to see the differences between how users manage virtualized workloads vs containerized workloads. Within the context of existing Kubernetes functionality, the best way to describe the VM controller’s behavior is to compare it to a StatefulSet of size one. This is because the VM controller represents a single stateful (immortal) virtual machine capable of persisting state across both node failures and multiple restarts of its underlying VMI. This object behaves in the way that is familiar to users who have managed virtual machines in AWS, GCE, OpenStack or any other similar IaaS cloud platform. The user can shutdown a VM, then choose to start that exact same VM up again at a later time.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: SIG Docs have some feelings against Latin in tech docs. :)
`vs` can be replaced with `, compared to`.
feel free to leave as is.

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+In addition to VMs, we also have a VirtualMachineInstanceReplicaSet (VMIRS) workload controller which manages scale out of identical VMI objects. This controller behaves nearly identically to the Kubernetes ReplicSet controller. The primary difference being that the VMIRS manages VMI objects and the ReplicaSet manages Pods. Wouldn’t it be nice if we could come up with a way to [use the Kubernetes ReplicaSet controller to scale out CRDs?](https://github.com/kubernetes/kubernetes/issues/65622)
+
+Each one of these KubeVirt objects (VMI, VM, VMIRS) are registered with Kubernetes as a CRD when the KubeVirt install manifest is posted to the cluster. By registering our apis as CRDs with Kubernetes, all the tooling involved with managing Kubernetes clusters (like kubectl) have access to the KubeVirt apis just as if they are native Kubernetes objects.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `apis` -> `APIs`.
in a couple of places in this paragraph.

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+## Dynamic Webhooks for API Validation
+
+One of the responsibilities of the Kubernetes apiserver is to intercept and validate requests prior to allowing objects to be persisted into etcd. For example If someone tries to create a pod using a malformed pod spec, the Kubernetes apiserver immediately catches the error and rejects the POST request. This all occurs before the object is persistent into etcd preventing the malformed pod spec from making its way into the cluster.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `Kubernetes apiserver` should be `Kubernetes API server`, unless its about `kube-apiserver`.
`For example If` -> `For example, if` (comma, lowercase if).
`pod` -> `Pod`.
`pod spec` -> `Pod specification`.
`Kubernetes apiserver` ->  `Kubernetes API server`
`pod spec` -> `Pod specification`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+This validation occurs during a process called admission control. Until recently, it was not possible to extend the default Kubernetes admission controllers without altering code and compiling/deploying an entirely new Kubernetes apiserver. This meant that if we wanted to perform admission control on KubeVirt’s CRD objects as they are posted to the cluster, we’d have to build our own version of the Kubernetes apiserver and convince people they should use it. That was not a viable solution for us.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `apiserver` ->  `API server`
`objects as they are` -> `objects, while they are` ?
`Kubernetes apiserver` -> `Kubernetes API server`
`convince people they should use it.` -> `convince our users to use that instead.` optional, feels less direct.

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+Using the new [Dynamic Admission Control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) feature in that first landed in Kubernetes 1.9 we now have a path for performing custom validation on KubeVirt api through the use of a [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#external-admission-webhooks). This feature allows KubeVirt to dynamically register an https webhook with Kubernetes at KubeVirt install time. After registering the custom webhook, all requests related to KubeVirt api objects are forwarded from the Kubernetes apiserver to our https endpoint for validation. If our endpoint rejects a request for any reason, the object will not be persisted into etcd and the client receives our response outlining the reason for the rejection.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `feature in` -> `feature`
`Kubernetes 1.9 we` -> `Kubernetes 1.9, we`
`api` -> `API`
`https webhook` -> `HTTPS webhook`
`KubeVirt api` -> `KubeVirt API`
`Kubernetes apiserver` ->  `Kubernetes API server`
`https endpoint` -> `HTTPS endpoint`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+A VMPreset acts in a similar way to a PodPreset. Just like a PodPreset allows users to define values that should automatically be injected into pods at creation time, a VMPreset allows users to define values that should be injected into VMs at creation time. Through the use of a mutating webhook, KubeVirt can intercept a request to create a VM, apply VMPresets to the VM spec, and then validate that the resulting VM object. This all occurs before the VM object is persisted into etcd which allows KubeVirt to immediately notify the user of any conflicts at the time the request is made.
+
+## Subresources for CRDs
+
+When comparing the use of CRDs to an aggregated API server, one of the features CRDs lack is the ability to support subresources. Subresources are used to provide additional resource functionality. For example, the ‘pod/logs’ and ‘pod/exec‘ subresource endpoints are used behind the scenes to provide the ‘kubectl logs’ and ‘kubectl exec’ command functionality.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: ‘pod/logs‘ -> `pod/logs`
‘pod/exec‘ -> `pod/exec`
‘kubectl logs’ -> `kubectl logs`
‘kubectl exec’ -> `kubectl exec`

^ added code formatting quotes - " ` "

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+Just like Kubernetes uses the pod/exec subresource to provide access to a pod’s environment, In KubeVirt we want subresources to provide serial-console, VNC, and SPICE access to a virtual machine. By adding virtual machine guest access through subresources, we can leverage RBAC to provide access control for these features.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `pod/exec`
`environment, In` -> `environment, in`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+So, given that the KubeVirt team decided to use CRD’s instead of an aggregated API server for custom resource support, how can we have subresources for CRDs when the CRD feature expiclity does not support subresources?
+
+We created a workaround for this limitation by implementing a stateless aggregated API server that exists only to serve subresource requests. With no state, we don’t have to worry about any of the issues we identified earlier with regards to access to etcd. This means the KubeVirt api is actually supported through a combination of both CRDs for resources and an aggregated API server for stateless subresources.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `KubeVirt api` -> `KubeVirt API`.

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+This isn’t a perfect solution for us. Both aggregated API servers and CRDs require us to register an api GroupName with Kubernetes. This api GroupName field essentially namespaces the api’s REST path in a way that prevents api naming conflicts between other third party applications. Because CRDs and aggregated API servers can’t share the same GroupName, we have to register two separate GroupNames. One is used by our CRDs and the other is used by the aggregated API server for subresource requests.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `api GroupName` -> `API GroupName` (2x)

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+We created a workaround for this limitation by implementing a stateless aggregated API server that exists only to serve subresource requests. With no state, we don’t have to worry about any of the issues we identified earlier with regards to access to etcd. This means the KubeVirt api is actually supported through a combination of both CRDs for resources and an aggregated API server for stateless subresources.
+
+This isn’t a perfect solution for us. Both aggregated API servers and CRDs require us to register an api GroupName with Kubernetes. This api GroupName field essentially namespaces the api’s REST path in a way that prevents api naming conflicts between other third party applications. Because CRDs and aggregated API servers can’t share the same GroupName, we have to register two separate GroupNames. One is used by our CRDs and the other is used by the aggregated API server for subresource requests. 
+
+Having two GroupNames in our api is slightly inconvenient because it means the REST path for the endpoints that serve the KubeVirt subresource requests have a slightly different base path than the resources.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `api` -> `API`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+**/apis/kubevirt.io/v1alpha2/namespaces/my-namespace/virtualmachineinstances/my-vm**
+
+However, the subresource endpoint to access graphical VNC looks like this.
+
+**/apis/subresources.kubevirt.io/v1alpha2/namespaces/my-namespace/virtualmachineinstances/my-vm/vnc**
+
+Notice that the first request uses **kubevirt.io** and the second request uses **subresource.kubevirt.io**. We don’t like that, but that’s how we’ve managed to combine CRDs with a stateless aggregated API server for subresources.
+
+One thing worth noting is that in Kubernetes 1.10 a very basic form of CRD subresource support was added in the form of the /status and /scale subresources. This support does not help us deliver the virtualization features we want subresources for. There have however been discussions about exposing custom CRD subresources as webhooks in a future Kubernetes version. If this functionality lands, we will gladly transition away from our stateless aggregated API server workaround to use a subresource webhook feature.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: /status and /scale -> `/status` and `/scale`
`There have however` -> `However, there have`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+## CRD Finalizers
+
+A [CRD finalizer](https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#advanced-topics) is a feature that lets us provide a pre-delete hook in order to perform actions before allowing a CRD object to be removed from persistent storage. In KubeVirt, we use finalizers to guarantee a virtual machine has completely terminated before we allow the corresponding VMI object to be removed from etcd. 
+
+## API Versioning for CRDs
+
+The Kubernetes core apis have the ability to support multiple versions for a single object type and perform conversions between those versions. This gives the Kubernetes core apis a path for advancing the v1alpha1 version of an object to a v1beta1 version and so forth.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `core apis` -> `core APIs`.
v1alpha1 -> `v1alpha1`
v1beta1 -> `v1beta1`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+Prior to Kubernetes 1.11, CRDs did not not have support for multiple versions. This meant when we wanted to progress a CRD from kubevirt.io/v1alpha1 to kubevirt.io/v1beta1, the only path available to was to backup our CRD objects, delete the registered CRD from Kubernetes, register a new CRD with the updated version, convert the backed up CRD objects to the new version, and finally post the migrated CRD objects back to the cluster.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: kubevirt.io/v1alpha1 to kubevirt.io/v1beta1 -> `kubevirt.io/v1alpha1` to `kubevirt.io/v1beta1`

File: content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+That strategy was not exactly a viable option for us.
+
+Fortunately thanks to some recent [work to rectify this issue in Kubernetes](https://github.com/kubernetes/features/issues/544), the latest Kubernetes v1.11 now supports [CRDs with multiple versions](https://github.com/kubernetes/kubernetes/pull/63830). Note however that this initial multi version support is limited. While a CRD can now have multiple versions, the feature does not currently contain a path for performing conversions between versions. In KubeVirt, the lack of conversion makes it difficult us to evolve our api as we progress versions. Luckily, support for conversions between versions is underway and we look forward to taking advantage of that feature once it lands in a future Kubernetes release.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-27-kubevirt-crds-for-virtualization.md]: `our api` -> `API`.

**Comment by davidvossel** [comment]
@neolit123 Thanks for the feedback. I believe I've addressed everything now.

**Pull request merged**